flink-test-utils-parent/flink-connector-testing/src/main/java/org/apache/flink/connectors/test/common/environment/MiniClusterTestEnvironment.java

```diff
@@ -128,6 +142,38 @@ private void startTaskManager() throws Exception {
         LOG.debug("New TaskManager {} has been launched.", latestTMIndex);
     }
 
+    @SuppressWarnings("unchecked")
+    private void waitUntilJobRestarted(JobClient jobClient, Deadline timeout) throws Exception {
+        checkNotNull(this.metricReporter, "In-memory metric reporter should exists");
+
+        // Get metric group of the running job
+        final Optional<JobManagerJobMetricGroup> jobMetricGroup =
+                metricReporter.findGroups("jobmanager").stream()
+                        .filter(group -> group instanceof JobManagerJobMetricGroup)
+                        .map(group -> (JobManagerJobMetricGroup) group)
+                        .filter(group -> group.jobId().equals(jobClient.getJobID()))
+                        .findFirst();
+        if (!jobMetricGroup.isPresent()) {
+            throw new IllegalStateException(
+                    String.format(
+                            "Cannot find metric group of the running job with ID %s",
+                            jobClient.getJobID()));
+        }
+
+        // Get numRestarts gauge
+        final Map<String, Metric> metrics = metricReporter.getMetricsByGroup(jobMetricGroup.get());
+        if (!metrics.containsKey(MetricNames.NUM_RESTARTS)) {
+            throw new IllegalStateException(
+                    String.format(
+                            "Cannot find metric \"%s\" in job metric group",
+                            MetricNames.NUM_RESTARTS));
+        }
+        final Gauge<Long> numRestarts = (Gauge<Long>) metrics.get(MetricNames.NUM_RESTARTS);
+
+        // Wait until the job is restarted at least once
+        CommonTestUtils.waitUntilCondition(() -> numRestarts.getValue() > 0, timeout);
```

Comment:
I'm not fond of the implementation because it depends a lot on the assumption that this method is only called once, and that previously the job has not restarted once for any other reason.